src/tools/genpolicy/src/layers_cache.rs

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2023 Microsoft Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use crate::registry::ImageLayer;
+
+use fs2::FileExt;
+use log::warn;
+use std::fs::OpenOptions;
+use std::sync::{Arc, Mutex};
+
+#[derive(Debug, Clone)]
+pub struct ImageLayersCache {
+    inner: Arc<Mutex<Vec<ImageLayer>>>,
+    filename: Option<String>,
+}
+
+impl ImageLayersCache {
+    pub fn new(layers_cache_file_path: &Option<String>) -> Self {
+        let layers = match ImageLayersCache::try_new(layers_cache_file_path) {
+            Ok(layers) => layers,
+            Err(e) => {
+                warn!("Could not read image layers cache: {e}");
+                Vec::new()
+            }
+        };
+        Self {
+            inner: Arc::new(Mutex::new(layers)),
+            filename: layers_cache_file_path.clone(),
+        }
+    }
+
+    fn try_new(layers_cache_file_path: &Option<String>) -> std::io::Result<Vec<ImageLayer>> {
+        match &layers_cache_file_path {
+            Some(filename) => {
+                let file = OpenOptions::new()
+                    .read(true)
+                    .write(true)
+                    .create(true)
+                    .truncate(false)
+                    .open(filename)?;
+                FileExt::lock_shared(&file)?;
```

Comment:
I vote for try_lock_shared - allowing this genpolicy instance to make progress even if another concurrent instance was stuck for some reason, and even if caching won't work in that case.